### DISCUSSION TITLE
Consultation regarding a replacement for rnodeconf - "rnodeman"

### DISCUSSION THREAD

**Discussion by jacobeva** [markqvist/Reticulum#550] [Ideas]
I'm currently in the drawing board stages of designing a replacement for rnodeconf written in C. For now, I've settled on the name "rnodeman".

I spoke some time ago about having the option to configure repositories for rnodeman in the Matrix channel. This has two purposes:
- to allow for downloading and installing RNode firmware (as to be expected)
- to allow for board recognition without baking it into the program

The second is quite important, as from the firmware side of things, adding additional board support is made a bit awkward as a PR needs to be made to rnodeconf so it can support the board each time, which is just no good. rnodeman would be able to refresh the list(s) of supported boards according to the repositories it points to in its config file.

The layout of the JSON file which would be downloadable from each repository would look as follows:
```
{
    "repo_name": "LES RNode FW CE",
    "boards":
    [
        { 
            "name": "RAK4631",
            "product": 16,
            "models": 
            [
                {
                    "model": 17,
                    "min_freq":430000000,
                    "max_freq":510000000,
                    "txp": 22,
                    "dl_dest": "230ffff37e83a28852058ca2f17036c9",
                    "dl_name": "rak4631_fw.zip",
                    "dl_fiat": "https://example.com/rak_fw_not_a_virus.zip",
                    "modems": [17, 33],
                    "lowest_fw_ver": 1.73,
                    "latest_fw_ver": 1.73
                }
            ]
        }
    ]
}

```
I should point out that product and model values are integer representations of the hex values currently configured in rnodeconf (JSON doesn't allow hex). Same with the modems. The idea would be to have the ability to download the firmware from a Reticulum destination and an internet destination (here referred to as dl_fiat, 1) because it's funny, but 2) also because it kind of makes sense).

Does anybody care to give input on the layout of this file? Any fields which may be needed which I have missed? I should also note the JSON will be able to be loaded locally, downloaded from an RNS destination, or over a fiat internet connection (it gets funnier every time).

**Comment by faragher** [comment]
Integer may cause confusion. Perhaps list hex as text strings and parse in the program?

**Reply by jacobeva** [reply]
Reviewing this now, it would be pretty easy to convert it from a hex string to long using `strtol()` in C, so I guess I'll update the spec.

**Comment by jacobeva** [comment]
New JSON spec:
```
{
    "repo_name": "LES RNode FW CE",
    "boards":
    [
        { 
            "name": "RAK4631",
            "product": "0x11",
            "models": 
            [
                {
                    "model": "0x12",
                    "min_freq":430000000,
                    "max_freq":510000000,
                    "txp": 22,
                    "dl_rns_dest": "230ffff37e83a28852058ca2f17036c9",
                    "dl_rns_path": "{ver}/rak4631_fw.zip",
                    "dl_fiat": "https://example.com/{ver}/rak_fw_not_a_virus.zip",
                    "modems": ["0x11"],
                    "lowest_ver": 1.73,
                    "latest_ver": 1.73
                }
            ]
        }
    ]
}
```

**Reply by jacobeva** [reply]
`{ver}` will be replaced with the version of the firmware `rnodeman` is requesting from the repo. The repo can define this string anywhere in the `dl_name` or `dl_fiat` params.

**Reply by jacobeva** [reply]
Note that the `dl_dest`, `dl_name` and `dl_fiat` params are all specified per model instead of per board. This is because some models require different firmware, e.g. the 433MHz version of the T-Echo (SX1268).

**Reply by jacobeva** [reply]
`dl_name` has been renamed to `dl_rns_path` for clarity. So has `dl_dest` to `dl_rns_dest`.
